File: 2021/csharp/Aoc2021/Day05.cs
namespace Aoc2021;

public class Day05 : IDay
{
    public override string Day { get; } = nameof(Day05)[3..];

    public override string Result1()
    {
        return CountMostDangerousPoints(InputLines, false)
            .ToString(CultureInfo.InvariantCulture);
    }

    public override string Result2()
    {
        return CountMostDangerousPoints(InputLines, true)
            .ToString(CultureInfo.InvariantCulture);
    }

    public static int CountMostDangerousPoints(IEnumerable<string> lines, bool withDiagonalLines)
    {
        return CountCoverage(
                lines
                .Select(Line.FromString)
                .Where(line => withDiagonalLines || !isDiagonalLine(line))
            )
            .Count(pointCount => pointCount.Value >= 2);

        static bool isDiagonalLine(Line line)
        {
            return line.P1.X != line.P2.X && line.P1.Y != line.P2.Y;
        }
    }

    private static Dictionary<Point, int> CountCoverage(IEnumerable<Line> lines)
    {
        return lines.Aggregate(new Dictionary<Point, int>(), addPoints);

        static Dictionary<Point, int> addPoints(Dictionary<Point, int> grid, Line line)
        {
            return line.Points().Aggregate(grid, incrementPoint);
        }

        static Dictionary<Point, int> incrementPoint(Dictionary<Point, int> grid, Point point)
        {
            _ = grid.TryAdd(point, 0);
            grid[point] += 1;
            return grid;
        }
    }

    public record Point(int X, int Y)
    {
        public static Point FromString(string input)
        {
            string[] chunks = input.Split(",", 2);
            return new(ParseInt(chunks[0]), ParseInt(chunks[1]));
        }

    }

    public record Line(Point P1, Point P2)
    {
        public IEnumerable<Point> Points()
        {
            int dx = Math.Sign(P2.X - P1.X);
            int dy = Math.Sign(P2.Y - P1.Y);
            int n = Math.Max(Math.Abs(P2.X - P1.X), Math.Abs(P2.Y - P1.Y)) + 1;
            for (int x = P1.X, y = P1.Y; n != 0; n--, x += dx, y += dy)
            {
                yield return new(x, y);
            }
        }

        public static Line FromString(string input)
        {
            string[] chunks = input.Split(" -> ", 2);
            return new(Point.FromString(chunks[0]), Point.FromString(chunks[1]));
        }
    }
}

File: 2021/fsharp/Aoc2021/Day05.fs
module Day05

open Utils

type Point = { X: int; Y: int }

module Point =
    let ofString string =
        let numbers = string |> String.split ',' |> Array.map int

        { X = numbers.[0]; Y = numbers.[1] }

type Line = { P1: Point; P2: Point }

module Line =
    let ofString (string: string) =
        let points = string.Split(" -> ") |> Array.map Point.ofString

        { P1 = points.[0]; P2 = points.[1] }

    open System

    let points { P1 = p1; P2 = p2 } =
        let dx = Math.Sign(p2.X - p1.X)
        let dy = Math.Sign(p2.Y - p1.Y)

        let n = Math.Max(Math.Abs(p2.X - p1.X), Math.Abs(p2.Y - p1.Y)) + 1

        seq {
            let mutable p = p1

            for _ = 1 to n do
                yield p
                p <- { X = p.X + dx; Y = p.Y + dy }
        }

type LineFilter =
    | WithDiagonals
    | WithoutDiagonals

let countMostDangerousPoints filter lines =
    let incrementPoint map point =
        map
        |> Map.change point (function
            | None -> 1 |> Some
            | Some n -> n + 1 |> Some)

    let addPoints map line =
        line |> Line.points |> Seq.fold incrementPoint map

    let isDiagonal line =
        line.P1.X <> line.P2.X && line.P1.Y <> line.P2.Y

    lines
    |> List.map Line.ofString
    |> List.filter (fun line ->
        match filter with
        | WithDiagonals -> true
        | WithoutDiagonals -> line |> isDiagonal |> not)
    |> List.fold addPoints Map.empty
    |> Map.values
    |> Seq.filter (fun value -> value >= 2)
    |> Seq.length

let input = "05" |> readInputLines |> Seq.toList

let job1 () =
    input |> countMostDangerousPoints WithoutDiagonals |> string

let job2 () =
    input |> countMostDangerousPoints WithDiagonals |> string
